e2e/LongHaul/device/IotHub.cs

```diff
@@ -135,23 +140,48 @@ public async Task SendTelemetryMessagesAsync(Logger logger, CancellationToken ct
                     continue;
                 }
 
-                // Make get a message to send, unless we're retrying a previous message
-                if (pendingMessage == null)
+                // Get messages to send, unless we're retrying a previous set of messages
+                if (!pendingMessages.Any())
                 {
-                    _messagesToSend.TryDequeue(out pendingMessage);
+                    for (int i = 0; i < maxBulkMessages; ++i)
+                    {
+                        if (sendSingle && pendingMessages.Count == 1)
+                        {
+                            break;
+                        }
+
+                        _messagesToSend.TryDequeue(out TelemetryMessage pendingMessage);
+                        if (pendingMessage == null)
+                        {
+                            break;
+                        }
+
+                        pendingMessages.Add(pendingMessage);
+                    }
                 }
 
                 // Send any message prepped to send
-                if (pendingMessage != null)
+                if (pendingMessages.Any())
                 {
-                    var sw = Stopwatch.StartNew();
-                    await _deviceClient.SendTelemetryAsync(pendingMessage, ct).ConfigureAwait(false);
+                    if (pendingMessages.Count > 1)
+                    {
+                        logger.Trace($"Sending {pendingMessages.Count} messages in bulk.");
+                        sw.Restart();
+                        await _deviceClient.SendTelemetryAsync(pendingMessages, ct).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        sw.Restart();
+                        await _deviceClient.SendTelemetryAsync(pendingMessages.First(), ct).ConfigureAwait(false);
+                    }
                     sw.Stop();
 
-                    _logger.Metric(TotalTelemetryMessagesSent, ++_totalTelemetryMessagesSent);
+                    _totalTelemetryMessagesSent += pendingMessages.Count;
+                    _logger.Metric(TotalTelemetryMessagesSent, _totalTelemetryMessagesSent);
                     _logger.Metric(TelemetryMessageDelaySeconds, sw.Elapsed.TotalSeconds);
 
-                    pendingMessage = null;
+                    sendSingle = !sendSingle;
```

Comment:
why flip `sendSingle` after sending all pending messages?

Reply:
To alternate between sending in bulk or single so AMQP gets both paths tested. I can add a code comment.

Reply:
got it. Sure, that could help!